modules/core/04-channel/keeper/upgrade.go

```diff
@@ -131,125 +96,98 @@ func (k Keeper) ChanUpgradeTry(ctx sdk.Context, portID string, channelID string,
 		)
 	}
 
-	if connectionEnd.GetCounterparty().GetConnectionID() != counterpartyUpgradeChannel.ConnectionHops[0] {
-		return 0, errorsmod.Wrapf(connectiontypes.ErrInvalidConnection, "unexpected counterparty channel connection hops, expected %s but got %s", connectionEnd.GetCounterparty().GetConnectionID(), counterpartyUpgradeChannel.ConnectionHops[0])
+	counterpartyConnectionHops := []string{connectionEnd.GetCounterparty().GetConnectionID()}
+	expectedCounterpartyChannel := types.Channel{
+		State:           types.INITUPGRADE,
+		Counterparty:    types.NewCounterparty(portID, channelID),
+		Ordering:        channel.Ordering,
+		ConnectionHops:  counterpartyConnectionHops,
+		Version:         channel.Version,
+		UpgradeSequence: counterpartyUpgradeSequence,
 	}
 
+	// verify that the counterparty channel has entered into the upgrade process
 	if err := k.connectionKeeper.VerifyChannelState(ctx, connectionEnd, proofHeight, proofChannel, channel.Counterparty.PortId,
-		channel.Counterparty.ChannelId, counterpartyUpgradeChannel); err != nil {
+		channel.Counterparty.ChannelId, expectedCounterpartyChannel); err != nil {
 		return 0, err
 	}
 
-	upgradeTimeout := types.UpgradeTimeout{TimeoutHeight: timeoutHeight, TimeoutTimestamp: timeoutTimestamp}
-	if err := k.connectionKeeper.VerifyChannelUpgradeTimeout(ctx, connectionEnd, proofHeight, proofUpgradeTimeout, channel.Counterparty.PortId,
-		channel.Counterparty.ChannelId, upgradeTimeout); err != nil {
+	// verify the proposed upgrade information provided by the counterparty
+	if err := k.connectionKeeper.VerifyChannelUpgrade(ctx, connectionEnd, proofHeight, proofUpgrade, channel.Counterparty.PortId,
+		channel.Counterparty.ChannelId, counterpartyProposedUpgrade); err != nil {
 		return 0, err
 	}
 
-	if err := k.connectionKeeper.VerifyChannelUpgradeSequence(ctx, connectionEnd, proofHeight, proofUpgradeSequence, channel.Counterparty.PortId,
-		channel.Counterparty.ChannelId, counterpartyUpgradeSequence); err != nil {
-		return 0, err
-	}
+	/*
+		replace the following with:
+
+		selfHeight := clienttypes.GetSelfHeight(ctx)
+		selfTime := uint64(ctx.BlockTime.UnixNano())
+		if info, hasPassed := counterpartyProposedUpgrade.Timeout.Status(selfHeight, selfTime); hasPassed {
+			return 0, types.NewErrorReceipt(upgradeSequence, info)
+		}
+	*/
 
 	// check if upgrade timed out by comparing it with the latest height of the chain
 	selfHeight := clienttypes.GetSelfHeight(ctx)
+	timeoutHeight := counterpartyProposedUpgrade.Timeout.TimeoutHeight
 	if !timeoutHeight.IsZero() && selfHeight.GTE(timeoutHeight) {
-		if err := k.RestoreChannelAndWriteErrorReceipt(ctx, portID, channelID, upgradeSequence, types.ErrUpgradeTimeout); err != nil {
+		if err := k.WriteErrorReceipt(ctx, portID, channelID, upgradeSequence, types.ErrUpgradeTimeout); err != nil {
```

Comment:
A lot simpler without the restore, but I would still like to write the error receipt in a single place in the final version. See #3418 as a reference.